### DISCUSSION TITLE
How would I install hpx with CMake alongside my application?

### DISCUSSION THREAD

**Discussion by McKillroy** [STEllAR-GROUP/hpx#5598] [Q&A] [answered]
I want to install the dependencies of my application alongside the binaries with CMake. 
What I'm trying is the CMake command `install(IMPORTED_RUNTIME_ARTIFACTS  <dependency targets list>)`. 
It works nicely with fmt, mimalloc, lua and others. 
HPX however is not added to my target with the usual target_link_libraries but with HPX_SETUP_TARGET.
My attempts to use HPX:hpx or HPX:wrap_main didn't work to get hpx.dll installed. 
The HPX::iostreams_component works. Which target could I use to install the hpx main library ?

Example cmake command:

    install(IMPORTED_RUNTIME_ARTIFACTS 
        fmt::fmt
        ${LUA_LIBRARIES}        
        mimalloc
        nng::nng
        #HPX::hpx #<-- Doesn't work: .. given target "HPX::hpx" which is not an executable, library, or module.		
        #HPX::wrap_main #<-- Doesn't work either
        HPX::iostreams_component  # <-- Works nicely
    )

**Comment by msimberg** [comment]
Have you tried *not* installing HPX explicitly at all? I'm assuming you're using fetchcontent (or similar) in your project, and HPX knows how to install itself (it issues those `install` commands for itself). I just tried the following minimal example:

```
cmake_minimum_required(VERSION 3.19)
project(myproject CXX)
include(FetchContent)
fetchcontent_declare(
  HPX
  GIT_REPOSITORY https://github.com/STEllAR-GROUP/hpx.git
  GIT_TAG master
)
fetchcontent_makeavailable(HPX)
add_executable(myexe main.cpp)
target_link_libraries(myexe PRIVATE HPX::hpx)
```
and building the `install` target builds and installs HPX.

As a side note, please consider using the CMake targets provided by HPX, not the macros. We're planning to remove the macros sooner or later. `HPX::hpx`, `HPX::wrap_main`, `HPX::component`, and `HPX::plugin` should cover all the use cases of the macro. If there's something missing that should be considered a bug. See here for more details: https://hpx-docs.stellar-group.org/latest/html/manual/creating_hpx_projects.html#using-cmake-targets.

**Reply by McKillroy** [reply]
Sorry, I was probably not explicit enough.
I am using vcpkg manifest mode to build my project and resolve all dependencies.
When running CMakes `INSTALL(TARGETS ...)` my application gets installed, but not the dependencies.
That's why I'm using `install(IMPORTED_RUNTIME_ARTIFACTS ...` to get the DLLs I need installed.

**Reply by msimberg** [reply]
Just to add a bit of context from the HPX side, the reason e.g. `HPX::hpx` and `HPX::wrap_main` can't be installed like the others is because they are interface targets and so they do not on their own have any imported artifacts, only transitively from other targets; `HPX::iostreams_component` is a plain `SHARED` CMake library and can be installed directly.

**Comment by McKillroy** [comment, marked as answer]
Finally found an experimental feature in vcpkg:
`set(X_VCPKG_APPLOCAL_DEPS_INSTALL ON)`
Also asked here: [https://github.com/microsoft/vcpkg/discussions/20526]
Also see [https://github.com/Microsoft/vcpkg/issues/1653]
